### DISCUSSION TITLE
Support for embedded animated GIF

### DISCUSSION THREAD

**Discussion by stoutgz** [pdf-association/pdf-issues#175] [Ideas]
I create documentation for engineers and create PDFs from markdown. The markdown original is hosted on Github in private repositories, but I convert the markdown to a PDF and share the PDF with individuals who don't have access to our Github repos. Sometimes animations are a great way to convey some UI behavior to an engineer. This works great on Github Markdown, but it would be really cool if I could also embed the GIF in a PDF so those with the PDF get a similar experience (be able to watch an animated GIF) to those viewing the web-rendered markdown. If there is already some way to do this, please enlighten me, otherwise, it would be great for the PDF standard to support embedded animations.

🙏🙏🙏

**Comment by petervwyatt** [comment]
My comments below reflect what the PDF specification permits in the _file format_. This may or may not be supported by different implementations and will vary across platforms (esp. mobile) - which are all specific vendor/implementation issues (in the same way that not all browsers support every corner of the HTML/CSS specs).

What you are talking about (animated GIFs) is really a simple form of stop motion animation, so there are various existing options:

* use of a Movie annotation (since PDF 1.2, but deprecated in PDF 2.0 in favour of other more flexible PDF annotation subtypes). Conversion of an animated GIF to a movie format should be easy. 

* use of a 3D annotation (since PDF 1.6) - this provides far more capabilities than trivial animation since a full interactive 3D experience is possible. I do note however that U3D native support in at least one well-known implementation was explicitly marked as "parsed but ignored". AFAIK PRC (the other 3D format supported in PDF 1.6) does not have "baked in" animation directly. Support for 3D PDF is widespread in the engineering communities with many CAD/ACE tools able to generate 3D PDFs (you mention this as your use-case).

* use of RichMedia annotations (since PDF 2.0) - this is the modern way of doing things as it provides more flexibility, more author control, less platform and implementation assumptions, better support for accessibility, isolating security concerns, etc. It also includes supports for U3D, PRC and STEP AP242 3D formats all of which are familiar to engineering communities, as well as other forms of rich multimedia (audio, movies, etc). Because it is newer and more flexible, implementation support still varies across vendors and across platforms (esp. mobile), with 3D PDF being the main industry driver so non-3D rich media support is likely to vary more. RichMedia annotations are the preferred method looking forward.

In all cases that use PDF annotations, the static appearance on the PDF page for printing, non-interactive processing and backwards compatibility is supported by the appearance stream of the relevant annotation. Appearance streams are now explicitly required in PDF.

Other concerns include ensuring that animations are made accessible to those with disabilities (so not autoplaying) and the security of the media runtime environment.

Then there are also Alternate Presentations (PDF 1.4, deprecated in PDF 2.0) but I wouldn't go this path.
